embeddings/pipeline/preprocessing_pipeline.py

```diff
@@ -52,120 +58,126 @@ def run(self) -> TransformationResult:
+# Mypy currently properly don't handle dataclasses with abstract methods  https://github.com/python/mypy/issues/5374
+@dataclass  # type: ignore
+class FlairPreprocessingPipeline(
+    PreprocessingPipeline[str, Union[datasets.DatasetDict, Corpus], Corpus],
+    abc.ABC,
 ):
-    def __init__(
-        self,
-        dataset_name_or_path: str,
-        persist_path: str,
-        input_column_name: str,
-        target_column_name: str,
-        datasets_path: Path = DATASET_PATH,
-        sample_missing_splits: Optional[Tuple[Optional[float], Optional[float]]] = None,
-        ignore_test_subset: bool = False,
-        seed: int = 441,
-        load_dataset_kwargs: Optional[Dict[str, Any]] = None,
-    ):
-        dataset = HuggingFaceDataset(
-            dataset_name_or_path, **load_dataset_kwargs if load_dataset_kwargs else {}
-        )
-        data_loader: Union[HuggingFaceDataLoader, HuggingFaceLocalDataLoader] = get_hf_dataloader(
-            dataset
-        )
-        transformation: Union[
-            Transformation[datasets.DatasetDict, Corpus], Transformation[Corpus, Corpus]
+    corpus_transformation_cls: Type[
+        Union[
+            ClassificationCorpusTransformation,
+            PairClassificationCorpusTransformation,
+            ColumnCorpusTransformation,
         ]
-        transformation = ClassificationCorpusTransformation(
-            input_column_name=input_column_name,
-            target_column_name=target_column_name,
-            datasets_path=datasets_path,
-        )
-        if sample_missing_splits:
-            transformation = transformation.then(
-                SampleSplitsFlairCorpusTransformation(*sample_missing_splits, seed=seed)
-            )
+    ] = field(init=False)
 
-        if ignore_test_subset:
-            transformation = transformation.then(DropSubsetFlairCorpusTransformation(subset="test"))
-        transformation = transformation.persisting(FlairPicklePersister(path=persist_path))
-        super().__init__(dataset, data_loader, transformation)
+    persister: Union[FlairConllPersister[Corpus], FlairPicklePersister[Corpus, Corpus]] = field(
+        init=False
+    )
 
+    dataset_name_or_path: str
+    persist_path: str
+    input_column_name: Union[str, Tuple[str, str]]
+    target_column_name: str
+    datasets_path: Path = DATASET_PATH
+    sample_missing_splits: Optional[Tuple[Optional[float], Optional[float]]] = None
+    downsample_splits: Optional[Tuple[Optional[float], Optional[float], Optional[float]]] = None
+    downsample_splits_stratification: bool = True
+    ignore_test_subset: bool = False
+    seed: int = 441
+    load_dataset_kwargs: Optional[Dict[str, Any]] = None
 
-class FlairTextPairClassificationPreprocessingPipeline(
-    PreprocessingPipeline[str, datasets.DatasetDict, Corpus]
-):
-    def __init__(
+    def __post_init__(self) -> None:
+        self._setup()
+        dataset, data_loader, transformation = self._setup_transformations()
```

Comment:
I dont like the idea that _setup_transformations() returns objects such as dataset, data_loader
It should only setup the transformation so in my understanding it shouldn't even return transformation but set self.transformation attribute